proposals/0000-replacing-cabal-custom-build.md

```diff
@@ -0,0 +1,1378 @@
+add a [`cabal doctest` external command using a simple script to invoke
+`doctest`](https://github.com/mpickering/cabal-doctest-demo).  This avoids the
+need for the package to use the `Custom` or `Hooks` build-type at all, which is
+generally preferable in most cases.
+
+In some cases
+(e.g. [`pretty-simple`](https://github.com/cdepillabout/pretty-simple/pull/129)),
+package authors may wish to integrate the doctests for their package into `cabal
+test`, in particular so that doctests can easily be executed by build tools
+other than `cabal-install`.  This will require the use of either
+`code-generators` or the `Hooks` build-type.
+
+#### Hooked programs
+
+The `hookedPrograms :: [Program]` field of `UserHooks` allows the custom
+`Setup.hs` file to specify new programs to be detected in the configure phase.
+In the `SetupHooks` design this use case is supported by having the
+`LocalBuildConfig` returned by the package-level pre-configure hook contain a
+field `withPrograms :: ProgramDb`, which can be extended by the hook.
+
+
+#### Composing `SetupHooks`
+
+A pattern observed in some `Setup.hs` code is that general-purpose `Cabal` build
+system extensions are implemented as functions of type `UserHooks -> UserHooks`,
+to allow for composition.
+
+For example, the [custom `Setup.hs` for
+`termonad`](https://hackage.haskell.org/package/termonad-4.5.0.0/src/Setup.hs)
+composes two such functions, one to add CPP options based on detecting system
+configuration, and the other to run doctests:
+
+```haskell
+main = do
+  cppOpts <- getTermonadCPPOpts
+  defaultMainWithHooks . addPkgConfigTermonadUserHook cppOpts $
+    addDoctestsUserHook "doctests" simpleUserHooks
+```
+
+To make composition of independent hooks more direct, we provide a `Monoid
+SetupHooks` instance, so this example
+[can become](https://gitlab.haskell.org/mpickering/hooks-setup-testing/-/blob/0575b7dd85d2151ec6ce4936c5d1800de869e356/patches/termonad-4.5.0.0.patch):
+
+```haskell
+setupHooks :: SetupHooks
+setupHooks = doctestsSetupHooks "doctests" <> termonadSetupHooks
+```
+
+In general the monoid is non-commutative, because for the
+pre-configure hooks the output of the first hook will be fed as an input to the
+second.
+
+
+## Alternatives
+
+### Effects available to hooks
+
+The proposal allows hooks to execute arbitrary IO effects. An alternative would
+be to limit the available effects using some kind of DSL, and thereby allow the
+possibility that the build tool could analyse the hooks or vary the
+implementation of the permitted effects.
+
+However, introducing a DSL would be a significant design contention point,
+because it is far from clear how best to design it or what effects are required.
+In general, the space of possible effects needed to augment the build system is
+unbounded (e.g. imagine a package that needs to generate code by parsing some
+binary file format, with a parser implemented using the FFI).
```

Comment:
This is why the "imagine all user code is loaded in one big process" though experiment I think that leads us astray.

Fundamentally build systems needs no effects at all. The whole thing is semantically a pure function, individual steps are pure functions.

The sandboxing / process-isolation mindset is "you, the guest rule, can think you are doing whatever crazy effects you want, but me the host do not care. I will fake all your effects; you are just a pure function to me with delusions."

In this case, sure a module needs to be generated by some code that needs to use a C library. But in practice it may well be better to force that to happen in a separate process, just for that one "generate module task:. The host build system should not be "sullied" by this FFI stuff. If we imagine it all in once process, we should be clear this is sort of an unsafe optimization.

What I worry is having the hooks file be Haskell is that we are putting the "unsafe optimization" cart before the "get the conceptual model right" horse.

It might be annoying for users to write separate exes for everything, but it makes it *very clear* what the host build system is supposed to do / be. The DSL is thus very simple, just a way to write a list of arguments, refer to a build-tool-depends and some `$<` / $@` variables to refer to input/outputs.